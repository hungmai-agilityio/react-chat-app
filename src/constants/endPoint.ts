export const URL = {
  HOME: '',
  SIGN_IN: '/sign-in',
  SIGN_UP: '/sign-up'
}

export const END_POINT = {
  USER: '/users',
  PROFILE: '/profiles',
  CHAT: 'chats',
  MESSAGE: 'messages'
}