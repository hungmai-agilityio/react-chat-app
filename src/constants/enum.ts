enum SIZE {
  MINI = 'extra-small',
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large'
}

enum TYPE {
  PRIMARY = 'primary',
  SECOND = 'second',
  TERTIARY = 'tertiary',
  FILE = 'file',
  TEXT = 'text',
  PASSWORD = 'password'
}

enum POSITION {
  TOP_RIGHT = 'top-right',
  TOP_LEFT = 'top-left',
  BOT_RIGHT = 'bottom-right',
  BOT_LEFT = 'bottom-left'
}

export { SIZE, TYPE, POSITION };
