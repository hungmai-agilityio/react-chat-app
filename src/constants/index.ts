export * from '@/constants/enum';
export * from '@/constants/list';
export * from '@/constants/endPoint';
export * from '@/constants/regex';
export * from '@/constants/message'
