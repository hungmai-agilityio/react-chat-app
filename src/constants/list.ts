import { ITab } from '@/components/Tabs/Tab';
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons';

export const USER_OPTIONS = [{ label: 'Profile', value: 'profile' }, { label: 'Log out', value: 'logout' }];

export const MESSAGE_OPTIONS = [
  { label: 'Edit', value: 'editMess', icon: faEdit },
  { label: 'Delete', value: 'deleteMess', icon: faTrash }
];

export const INFO_OPTIONS = [{ label: 'Chat info', value: 'infoChat' }];

export const LIST_TAB_USERS: Omit<ITab, 'index' | 'onClick'>[] = [
  {
    title: 'Chat',
    value: '1'
  },
  {
    title: 'Users',
    value: '2'
  }
];