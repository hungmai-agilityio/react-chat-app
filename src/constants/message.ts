export const MESSAGE_VALID = {
  INPUT_REQUIRED: '{field} is required',
  EMAIL_ERROR: 'Email is not valid',
  EMAIL_EXIST: 'Email has been registered, please try again!',
  PASSWORD_ERROR:
    'Minimum 6 characters, one uppercase, one lowercase, one number',
  CONFIRM_ERROR: 'Password confirmation do not match password',
};

export const MESSAGE_API = {
  SIGN_UP_ERROR: 'An error occurred. Please try again',
  SIGN_IN_ERROR: 'email or password is incorrect',
  UPDATE_PROFILE_SUCCESS: 'Update success',
  UPDATE_PROFILE_ERROR: 'Update error',
};