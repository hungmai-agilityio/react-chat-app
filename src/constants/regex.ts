const EMAIL_REGEX = /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/;
const PASSWORD_REGEX = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{6,}$/;

export { EMAIL_REGEX, PASSWORD_REGEX };
