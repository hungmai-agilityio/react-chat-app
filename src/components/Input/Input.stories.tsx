import { Meta, StoryObj } from '@storybook/react';

// Components
import Input from '@/components/Input';

// Constants
import { TYPE } from '@/constants';

// FontAwesome
import { faUser } from '@fortawesome/free-solid-svg-icons';

const meta: Meta<typeof Input> = {
  component: Input,
  tags: ['autodocs'],
  argTypes: {
    name: {
      description: 'Set input name'
    },
    label: {
      description: 'label tag name'
    },
    variant: {
      description: 'Set type optional for Input'
    },
    onClick: {
      description: 'Handle the event when clicking on the icon'
    },
    styles: {
      description: 'Add custom css'
    },
    disabled: {
      description: 'Set disabled input'
    },
    icon: {
      description: 'Render icon with FontAwesome'
    },
    message: {
      description: 'Render error message'
    },
    htmlFor: {
      description: 'Set htmlFor and input id'
    },
    placeholder: {
      description: 'Placeholder when not input value'
    },
    type: {
      description:
        'Type input, default is "text", optional(email, password, date, ...)'
    },
    value: {
      description: 'Input value'
    },
    onChange: {
      description: 'Handle change input value'
    },
    onKeyDown: {
      description: 'Handle keyboard click events'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Input>;

export const Primary: Story = {
  args: {
    variant: TYPE.PRIMARY,
    name: 'primaryInput',
    placeholder: 'Primary input',
    onChange: () => {},
    value: ''
  }
};

export const Secondary: Story = {
  args: {
    variant: TYPE.SECOND,
    name: 'secondaryInput',
    placeholder: 'Secondary input',
    onChange: () => {},
    value: ''
  }
};

export const Tertiary: Story = {
  args: {
    variant: TYPE.TERTIARY,
    name: 'tertiaryInput',
    placeholder: 'Tertiary input',
    onChange: () => {},
    value: ''
  }
};

export const InputWithIcon: Story = {
  args: {
    variant: TYPE.PRIMARY,
    name: 'inputWithIcon',
    placeholder: 'Input with icon',
    onChange: () => {},
    value: '',
    icon: faUser,
    onClick: () => alert('Icon clicked')
  }
};

export const InputWithMessage: Story = {
  args: {
    variant: TYPE.PRIMARY,
    name: 'inputWithMessage',
    placeholder: 'Input with message',
    onChange: () => {},
    value: '',
    message: 'This is an error message'
  }
};

export const InputLabel: Story = {
  args: {
    variant: TYPE.PRIMARY,
    name: 'largeInput',
    placeholder: 'Large input',
    onChange: () => {},
    value: '',
    label: 'label'
  }
};
