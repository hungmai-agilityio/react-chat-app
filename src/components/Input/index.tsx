import { ChangeEvent, useRef } from 'react';
import { clsx } from 'clsx';

// Font Awesome
import { IconDefinition } from '@fortawesome/free-solid-svg-icons';

// Constants
import { TYPE } from '@/constants';

// Components
import MemoizedIcon from '@/components/Button/Icon';

interface InputProps {
  label?: string;
  value?: string;
  variant?: TYPE.PRIMARY | TYPE.SECOND | TYPE.TERTIARY;
  htmlFor?: string;
  placeholder?: string;
  name?: string;
  onChange?: (
    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  onClick?: () => void;
  message?: string;
  icon?: IconDefinition;
  styles?: string;
  isDisabled?: boolean;
  type?: string;
}

const Input = ({
  label,
  value,
  variant,
  htmlFor,
  placeholder,
  name,
  onChange,
  onClick,
  message,
  icon,
  styles,
  isDisabled,
  type = TYPE.TEXT
}: InputProps) => {
  const inputRef = useRef<HTMLInputElement>(null);
  const textareaRef = useRef<HTMLTextAreaElement>(null);

  const typeClasses = {
    [TYPE.PRIMARY]: 'h-11 border border-gray-400 rounded-md',
    [TYPE.SECOND]: 'h-10 border border-gray-400 focus:h-20 resize-none',
    [TYPE.TERTIARY]: 'h-10 border-b border-gray-400 text-center'
  };

  const commonClasses = 'w-full mt-1.5 outline-none p-3 focus:border-primary';
  const typeClass = typeClasses[variant || TYPE.PRIMARY];

  return (
    <div className="mb-3">
      {label && (
        <label htmlFor={htmlFor} className="font-bold leading-2-xs capitalize">
          {label}
        </label>
      )}
      <div className="relative">
        {variant === TYPE.SECOND ? (
          <textarea
            ref={textareaRef}
            id={htmlFor}
            name={name}
            value={value}
            onChange={onChange}
            placeholder={placeholder}
            className={clsx(
              'py-1 px-2 focus:h-20',
              commonClasses,
              typeClass,
              styles
            )}
            disabled={isDisabled}
          />
        ) : (
          <input
            ref={inputRef}
            id={htmlFor}
            name={name}
            value={value}
            onChange={onChange}
            type={type}
            placeholder={placeholder}
            className={clsx(commonClasses, typeClass, styles)}
            disabled={isDisabled}
          />
        )}
        {icon && (
          <MemoizedIcon
            icon={icon}
            onClick={onClick}
            styles="absolute top-2 right-2"
          />
        )}
      </div>
      {message && (
        <span className="text-red-500 mr-1 visible text-xs">{message}</span>
      )}
    </div>
  );
};

export default Input;
