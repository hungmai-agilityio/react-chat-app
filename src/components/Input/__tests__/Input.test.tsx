import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

// FontAwesome
import { faPen } from '@fortawesome/free-solid-svg-icons';

// Constants
import { TYPE } from '@/constants';

// Components
import Input from '@/components/Input';

// Test Data
const propsDefault = {
  label: 'Label',
  value: 'test value',
  htmlFor: 'input-id',
  placeholder: 'Enter text',
  name: 'input-name',
  onChange: jest.fn(),
  onClick: jest.fn(),
  message: '',
  icon: faPen,
  styles: '',
  isDisabled: false
};

// Test Cases
describe('Input', () => {
  it('renders PRIMARY type input', () => {
    render(<Input {...propsDefault} variant={TYPE.PRIMARY} />);

    const inputElement = screen.getByDisplayValue('test value');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveClass('h-11 border border-gray-400 rounded-md');
  });

  it('renders SECOND type textarea', () => {
    render(<Input {...propsDefault} variant={TYPE.SECOND} />);
    const textareaElement = screen.getByDisplayValue(
      'test value'
    ) as HTMLTextAreaElement;

    expect(textareaElement).toBeInTheDocument();
    expect(textareaElement).toHaveClass(
      'h-10 border border-gray-400 focus:h-20 resize-none'
    );
  });

  it('renders TERTIARY type input', () => {
    render(<Input {...propsDefault} variant={TYPE.TERTIARY} />);

    const inputElement = screen.getByDisplayValue('test value');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveClass(
      'h-10 border-b border-gray-400 text-center'
    );
  });

  it('renders with default type when no type is provided', () => {
    render(<Input {...propsDefault} variant={undefined} />);

    const inputElement = screen.getByDisplayValue('test value');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveClass('h-11 border border-gray-400 rounded-md');
  });

  it('shows message when provided', () => {
    render(<Input {...propsDefault} message="This is an error message" />);

    const messageElement = screen.getByText('This is an error message');
    expect(messageElement).toBeInTheDocument();
  });

  it('should render match to snapshot', () => {
    const { container } = render(<Input {...propsDefault} />);
    expect(container).toMatchSnapshot();
  });
});
