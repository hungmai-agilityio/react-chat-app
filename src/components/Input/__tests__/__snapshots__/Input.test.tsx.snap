// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render match to snapshot 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="font-bold leading-2-xs capitalize"
      for="input-id"
    >
      Label
    </label>
    <div
      class="relative"
    >
      <input
        class="w-full mt-1.5 outline-none p-3 focus:border-primary h-11 border border-gray-400 rounded-md"
        id="input-id"
        name="input-name"
        placeholder="Enter text"
        type="text"
        value="test value"
      />
      <div
        class="flex items-center justify-center cursor-pointer text-gray-400 h-10 w-10 rounded-full hover:text-dark hover:bg-tertiary absolute top-2 right-2"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pen "
          data-icon="pen"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M362.7 19.3L314.3 67.7 444.3 197.7l48.4-48.4c25-25 25-65.5 0-90.5L453.3 19.3c-25-25-65.5-25-90.5 0zm-71 71L58.6 323.5c-10.4 10.4-18 23.3-22.2 37.4L1 481.2C-1.5 489.7 .8 498.8 7 505s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L421.7 220.3 291.7 90.3z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
